docs/http-cloud-provider.md

```diff
@@ -0,0 +1,643 @@
+|ingress:ip|The IP address of the ingress point. This is a optional field as only one of IP or hostname is required.|"127.0.0.1"|
+|ingress:hostname|The hostname of the ingress point. This is a optional field as only one of IP or hostname is required.|"my-cloud-host1"|
+|exists|Whether the TCP Load Balancer exists or not.|true|
+Example
+
+```json
+{
+    "ingress": [
+        {
+            "ip":"127.0.0.1",
+            "hostname":"my-cloud-host1"
+        }
+    ]
+}
+```
+
+#### PUT /v1/tcpLoadBalancer/{region}/{name}
+
+Updates the hosts given in the Load Balancer specified.
+
+##### Sender body:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|hosts|A list of all host names.||
+|hosts:hostname|A host name.|"my-cloud-host1"|
+Example
+
+```json
+{
+    "hosts":[
+        {
+            "hostname":"my-cloud-host1"
+        }
+    ]
+}
+```
+
+##### Expected return body:
+
+Send back an empty 204 response if the object was updated successfully.
+
+#### DELETE /v1/tcpLoadBalancer/{region}/{name}
+
+Deletes the specified Load Balancer.
+
+##### Expected return value:
+
+Send back an empty 204 response if the object was deleted successfully or it didn't exist before.
+
+## Zones
+
+The Zones interface allows the kubernetes platform to query the cloud provider about the Zones and which areas are having a failed state.
+
+### Methods
+
+#### GET /v1/zones/supported
+
+Know whether the interface is supported or not.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|supported|Whether the interface is supported or not.|true|
+Example
+
+```json
+{
+    "supported":true
+}
+```
+
+#### GET /v1/zones
+
+Returns the Zone containing the current failure zone and locality region that the program is running in.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|failureDomain|The failure zone.|"my-zone-2"|
+|region|Locality region that the program is running in.|"USA"|
+Example
+
+```json
+{
+    "failureDomain":"my-zone-2",
+    "region":"USA"
+}
+```
+
+## Clusters
+
+The clusters interface allows the kubernetes platform to query information about the clusters inside the cloud provider.
+
+### Methods
+
+#### GET /v1/clusters/supported
+
+Know whether the interface is supported or not.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|supported|Whether the interface is supported or not.|true|
+Example
+
+```json
+{
+    "supported":true
+}
+```
+
+#### GET /v1/clusters
+
+Returns a list of all clusters.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|clusters|The list of all cluster objects.||
+|clusters:clusterName|The name of the cluster.|"my-cluster"|
+Example
+
+```json
+{
+    "clusters": [
+        {
+            "clusterName":"my-cluster"
+        }
+    ]
+}
+```
+
+#### GET /v1/clusters/{clusterName}/master
```

Comment:
It makes me unhappy that this is method is called Master in the cloudprovider interface. We should change it to API or something.

https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L49